MP REST Client config

TomasHofman opened the pull request:
This exposes following configuration options via the `application.properties` file:

```
org.acme.rest.client.CountriesService/mp-rest/connectionTTL
org.acme.rest.client.CountriesService/mp-rest/maxPooledPerRoute
org.acme.rest.client.CountriesService/mp-rest/connectionCheckoutTimeout
org.acme.rest.client.CountriesService/mp-rest/connectionPoolSize
```

(Assume org.acme.rest.client.CountriesService is an MP rest client interface.)

This could help with connection pooling issues like https://github.com/quarkusio/quarkus/issues/15383.


AFAIK the only other way to set above properties is via registering a `RestClientBuilderListener` like the one bellow, but I didn't find it documented anywhere. Being able to configure those via `application.properties` would be more developer friendly.

```
public class CustomRestClientBuilderListener implements RestClientBuilderListener {

    @Override
    public void onNewBuilder(RestClientBuilder builder) {
        builder.property("resteasy.connectionTTL", Arrays.asList(2345L, TimeUnit.MILLISECONDS));
        builder.property("resteasy.maxPooledPerRoute", 1);
    }
}
```

gsmet commented:
@michalszynkiewicz could you have a look at this one? It looks like a worthwhile addition.

kenfinnigan commented:
Not sure what @radcortez thinks, my only concern would be around the namespacing with the `/mp-rest/` prefix to the values.

Configuration off that prefix is defined in the MicroProfile REST Client spec. As what is being proposed is not part of the spec, we should probably include them in a different way.

TomasHofman commented:
Thanks for feedback. Would `.../restclient/...` make more sense then?

gsmet commented:
That is going to be very confusing.

I always thought we should have used proper Quarkus config for this.

radcortez commented:
> That is going to be very confusing.
> 
> I always thought we should have used proper Quarkus config for this.

Hum... how about if we also use `quarkus`? The `mp-rest` subnamespace is reserved for MP only config to not clash with specific runtime config and we can add what we want into the `quarkus` one. Of course, we can also expose all the current `mp-rest` configuration in the `quarkus` one.

So in essense:
- `mp-rest` only MP standard configuration
- `quarkus` every `mp-rest` configuration, plus specific things

May probably be worthwhile to define a specific vendor namespace in the MP Rest Client spec.

kenfinnigan commented:
I agree it makes sense to have a configuration key that isn't tied to the current MP way of defining REST Client config

TomasHofman commented:
@radcortez do I understand correctly that we would keep the class name prefix for both options?

So we would have:

```
org.acme.rest.client.CountriesService/mp-rest/url=http://localhost:8080/
# etc...
```

and then also:

```
org.acme.rest.client.CountriesService/quarkus/url=http://localhost:8080/
org.acme.rest.client.CountriesService/quarkus/connectionTTL=2000
org.acme.rest.client.CountriesService/quarkus/maxPooledPerRoute=3
# etc...
```

radcortez commented:
Yes, that is the proposal. We probably need a few +1 before actually going for it. @michalszynkiewicz what do you think?

kenfinnigan commented:
We'd duplicate the `url`?

TomasHofman commented:
From @radcortez's message:

> So in essense:
> *   mp-rest only MP standard configuration
> *   quarkus every mp-rest configuration, plus specific things

I gather the `quarkus` namespace would expose everything that `mp-rest` exposes? :)

kenfinnigan commented:
Wouldn't be confusing for Quarkus to duplicate all the MP specific config?

radcortez commented:
We can just relocate the configuration and have `quarkus` to take priority and fallback to `mp-rest`.

I'm not sure which one is more confusing. If having to use two namespaces, or use `quarkus` for everything. From my perspective, I think that if you could set everything on `quarkus` it would be cleaner. You don't have to guess which property goes where.

But I would like to hear other thoughts.

xstefank commented:
I agree that duplicating is probably not a good idea. Just keep the custom properties under the different namespace. Why would be:

radcortez commented:
I think we all agree on the new properties going to the Quarkus namespace, just not about the duplication. We can probably more like this and duplicate them later if needed.

BTW, another thought I had about this is, when we do an implementation first, and configurations get moved into the spec, they are renamed, but the implementation always retains the original names. In here we are doing the other way around. So what happens if one of the new configurations ends up in the specification? We are going to have spec only configurations, like we have now, quarkus only configurations, and maybe some that are on both spec and quarkus namespaces, but you will have to guess which ones.

kenfinnigan commented:
Another aspect is that we need to replicate these new configurations in the RESTEasy Reactive Client, not just the original client.

michalszynkiewicz commented:
What's the reason for creating a separate namespace than `mp-rest`? Are we afraid that at some point properties with the same name will be introduced in MicroProfile Rest Client?

TomasHofman commented:
> Another aspect is that we need to replicate these new configurations in the RESTEasy Reactive Client, not just the original client.

Thanks for pointing this out. I briefly looked at the reactive client and it looks it tries to appear the same to developer (do everything as you would with the classic rest client), but naturally it uses different tech under the hood (vertx). The vertx HTTP client offers different configuration options from the Resteasy one. E.g. there is no "connectionTTL" property in vertx, but there is the "keepAliveTimeout" property, which I assume provides similar functionality (but I didn't have time to test it yet).

Now there the question how to approach this:

1. Provide unified configuration options across both clients, which could be problematic if we want to expose something which is only available in one of the implementations.
2. Provide different config options for each client, which might ask for separate config namespaces?

From my brief investigation (I might be wrong), only two of the four config options I wanted to expose have equivalent in VertX:

connectionTTL ~ keepAliveTimeout
maxPooledPerRoute ~ no equivalent in vertx
connectionCheckoutTimeout ~ no equivalent in vertx
connectionPoolSize ~ maxPoolSize

The only "very important" attribute of those four is "connectionTTL" / "keepAliveTimeout" (because it's needed to overcome the problem with load balancers/proxies using a shorter timeout), the others are just hitching a ride. So if I wanted to simplify as much as possible, I'm OK with exposing this one, call it "connectionTTL" for both clients and use the same namespace. This might work until we realize another attribute is needed.

TomasHofman commented:
There's another suggestion from @stuartwdouglas on google groups:

> What about something like:
>
> * org.acme.rest.client.CountriesService/mp-rest/quarkus/connectionTTL
> * org.acme.rest.client.CountriesService/mp-rest/quarkus/maxPooledPerRoute
> * org.acme.rest.client.CountriesService/mp-rest/quarkus/connectionCheckoutTimeout
> * org.acme.rest.client.CountriesService/mp-rest/quarkus/connectionPoolSize
>
> This way it is similar to the existing config, but won't conflict with any future properties that may be defined.

radcortez commented:
> What's the reason for creating a separate namespace than `mp-rest`? Are we afraid that at some point properties with the same name (and different meaning) will be introduced in MicroProfile Rest Client?

Yes. I agree the chances for it to happen are slim. Maybe we are just adding complexity when there is none.

> 
> If we want to have a separate thing from `mp-rest`, I don't like `quarkus` for the name of the namespace. For me it should be something like `rest-client`, `rest`, `quarkus-rest`; `quarkus-rest-client` if we want to be really verbose.

Using `rest-client` sounds good. No references to any particular API or implementation.

> 
> If we're going to have custom properties under a new namespace, I'd be in favor of having all the properties under the new namespace, and quarkus-specific namespace always winning with MicroProfile ones if duplicated. IMO, we should then advertise the new namespace only. Pro: we get a consistent way of configuring clients, con: documenting MicroProfile stuff becomes a bit more tricky.

+1

> 
> I'm really not convinced having a separate namespace is the way to go. But if we go this way, and advertise it for all properties, it may shield us from potential namespace change when/if MicroProfile Rest Client moves to Jakarta Rest

So do you think we should just put any new things in `mp-rest`?

stuartwdouglas commented:
We should definitely not have split namespaces, were some stuff is under MP config and some stuff is under `quarkus`, this is super confusing. IMHO this is already pretty verbose as this has to be defined for every client interface, when often they will all be pointing at the same place. If we are going to offer quarkus options then IMHO we should do something like:
```
quarkus.rest-client.url=http://localhost:8080  #The default URL for all clients
quarkus.rest-client.beerservice.url=http://beer.com #the URL for 'beerservice' clients, as determined by an annotation
```

The beerservice would be something like:

```
@ClientGroup("beerservice")
public interface RatingService { ...
```

or a meta annotation:

```
@ClientGroup("beerservice")
public @interface BeerService {}

@BeerService
public interface RatingService {
```

This way we are not just duplicating the MP config, but are offering something that is IMHO much more useful and user friendly.

michalszynkiewicz commented:
@radcortez yes, I might be wrong but it seems to be the solution with least drawbacks to me.

@stuartwdouglas what's your opinion on adding the new properties under `mp-rest`?

I like the idea of trying to do better than the spec. I'm not sure though if `@ClientGroup` name wouldn't be a bit confusing for the users. I think using a single client for one base url is the most common case.

Side note: grouping might work (although this wasn't probably designed for it) in MP Rest Client with:

```java
@RegisterRestClient(configKey="myClient")
public interface MyClient1 {
}

@RegisterRestClient(configKey="myClient")
public interface MyClient1 {
}
```

configured by:
```properties
myClient/mp-rest/url=http://beer.com
```

stuartwdouglas commented:
The only issue I have with adding new properties under mp-rest is that there is potential for conflicts, but that is probably not such a big deal.

I think for some of these settings we should be able to set global defaults as well.

TomasHofman commented:
Taking into account the input presented here and also a related issue https://github.com/quarkusio/quarkus/issues/17832 (which calls for providing alternative way of configuring rest client and GraphQL, that wouldn't use slash character in config properties), I'm proposing following solution:

* The current config schema `ClassName/mp-rest/property-name` would remain in place with no changes (no new properties).
* A new schema would be created as proposed in the https://github.com/quarkusio/quarkus/issues/17832, that would be consistent with the rest of Quarkus configuration. It can look like this:
```
# configuration via config key:
quarkus.rest-client.countries-service.url=https://restcountries.eu/rest
quarkus.rest-client.countries-service.scope=Singleton
# configuration via class name:
quarkus.rest-client."org.acme.rest.client.CountriesService".url=https://restcountries.eu/rest
quarkus.rest-client."org.acme.rest.client.CountriesService".scope=Singleton
```
* Two new properties, `connectionTTL` and `connectionPoolSize` will be introduced into the new schema. These will be working for both the classic and reactive clients.
* The new schema will override properties set via the old one.

This allows for grouping clients (multiple client interfaces sharing the same config) by defining the same config key in the client interfaces: `@RegisterRestClient(configKey="country-service")`.

TomasHofman committed:
Add quarkus config properties for MP REST Client - rest client reactive part

TomasHofman committed:
Add quarkus config properties for MP REST Client - rest client classic part

TomasHofman commented:
Hello, I updated the PR.

The original (MP standard) configuration schema remains unchanged (except for couple of additions related to MP Rest Client 2.0, more on that bellow). Parallel to that, a new Quarkus style configuration schema was added, which exposes two extra attributes - `connectionTTL` and `connectionPoolSize`.

The new schema (under `quarkus.rest-client.*`), including the two new attributes, works for both classic and reactive REST client. If both MP style and Quarkus style property is configured, Quarkus one takes precedence.

There is however couple of non-standard properties that were already implemented in the reactive client, which remain functional only in the reactive client and have no effect on the classic client. I added these to the new schema too, but perhaps it would be better to put these into a separate reactive client schema? These attributes are (original -> new name):

* `quarkus.rest-client-reactive.disable-smart-produces` -> `quarkus.rest-client.disable-smart-produces`,
* `quarkus.rest.client.multipart-post-encoder-mode` -> `quarkus.rest-client.multipart-post-encoder-mode`,
* `quarkus.rest.client.max-redirects` -> `quarkus.rest-client.<client-prefix>.max-redirects`.


Some listings bellow...

### Original schema

The MP standard props:

```
<prefix>/mp-rest/url=
<prefix>/mp-rest/uri=
<prefix>/mp-rest/scope=
<prefix>/mp-rest/providers=
<prefix>/mp-rest/connectTimeout=
<prefix>/mp-rest/readTimeout=
<prefix>/mp-rest/hostnameVerifier=
<prefix>/mp-rest/trustStore=
<prefix>/mp-rest/trustStorePassword=
<prefix>/mp-rest/trustStoreType=
<prefix>/mp-rest/keyStore=
<prefix>/mp-rest/keyStorePassword=
<prefix>/mp-rest/keyStoreType=

<prefix>/mp-rest/followRedirects=
<prefix>/mp-rest/proxyAddress=
<prefix>/mp-rest/queryParamStyle=
```

The last three were implemented only in the reactive client, and now are functional in the classic client too.

Following non-standard properties were already implemented in the reactive client:

```
quarkus.rest.client.max-redirects=
quarkus.rest.client.multipart-post-encoder-mode=
quarkus.rest-client-reactive.disable-smart-produces=
```

### New schema

```
quarkus.rest-client."client-prefix".url=
quarkus.rest-client."client-prefix".uri=
quarkus.rest-client."client-prefix".scope=
quarkus.rest-client."client-prefix".providers=
quarkus.rest-client."client-prefix".connect-timeout=
quarkus.rest-client."client-prefix".read-timeout=
quarkus.rest-client."client-prefix".hostname-verifier=
quarkus.rest-client."client-prefix".trust-store=
quarkus.rest-client."client-prefix".trust-store-password=
quarkus.rest-client."client-prefix".trust-store-type=
quarkus.rest-client."client-prefix".key-store=
quarkus.rest-client."client-prefix".key-store-password=
quarkus.rest-client."client-prefix".key-store-type=
quarkus.rest-client."client-prefix".follow-redirects=true
quarkus.rest-client."client-prefix".proxy-address=
quarkus.rest-client."client-prefix".query-param-style=

# non-standard, works in both reactive and classic:
quarkus.rest-client."client-prefix".connection-ttl=
quarkus.rest-client."client-prefix".connection-pool-size=

# non-standard, works in reactive only - maybe move to separate reactive schema?
quarkus.rest-client."client-prefix".max-redirects=
quarkus.rest-client.disable-smart-produces=
quarkus.rest-client.multipart-post-encoder-mode=
```

TomasHofman commented:
FTR there are legitimate native test failures. I will investigate.

FroMage commented on extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientConfig.java:
```diff
@@ -0,0 +1,212 @@
+
+    /**
+     * The CDI scope to use for injection. This property can contain either a fully qualified class name of a CDI scope
+     * annotation (such as "javax.enterprise.context.ApplicationScoped") or its' simple name (such as
```
```suggestion
     * annotation (such as "javax.enterprise.context.ApplicationScoped") or its simple name (such as
```

FroMage commented on extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilder.java:
```diff
@@ -316,4 +365,47 @@ private void configureBaseUrl(RestClientBuilder builder) {
     }
 
+    private <PropertyType> Optional<PropertyType> getClientConfigPropertyWithFallback(
+            Function<RestClientConfig, Optional<PropertyType>> method, String propertyFormat, Class<PropertyType> type) {
```
This looks like we're using a lot of lambdas just for config. Is it really reasonable? We're trying to limit the number of lambdas/classes for better perf.

TomasHofman replied:
This was fixed also due to @radcortez suggestion to use MP Config API to handle backward compatibility. No method references any more.

FroMage reviewed (changes requested):
This LGTM, except the question around possible excessive lambda usage. And we need to verify why the tests don't pass. Also, question: did you really implement that `@RegisterRestClient.configKey` defaults to the simple class name? 

`RestClientConfigRoot` says:

```
The key can be either the value of the configKey parameter of a `@RegisterRestClient` annotation, or the name of a class bearing that annotation, in which case it is possible to use the short name, as well as fully qualified name.
```

But I didn't see this tested, is it? Do we allow `quarkus.rest-client.EchoClient` as well as `quarkus.rest-client.echo-client`?

radcortez commented on extensions/resteasy-classic/rest-client/pom.xml:
```diff
@@ -16,5 +16,6 @@
     <modules>
         <module>deployment</module>
         <module>runtime</module>
+        <module>config</module>
```
Why is this a separate module? It doesn't follow our usual model for extensions.

TomasHofman replied:
@radcortez I want to use the configuration classes in both the classic and reactive rest client extensions, so I need some module that can be consumed by both. I don't see any suitable module that would be already shared by the two client extensions.

Should I create a `resteasy-classic/rest-client-config/runtime/` structure, or something else entirely? I don't want to change the current directory structure too much...

radcortez commented on extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientConfigRoot.java:
```diff
@@ -0,0 +1,55 @@
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "rest-client", phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
+public class RestClientConfigRoot {
```
`RestClientConfig` seems more suitable as a name for a public API if we expect users to inject it directly into their services. Usually extension configuration roots are not used outside extensions so the name doesn't matter much. In this case, we may want to be careful about that.

TomasHofman replied:
I'm going to change the root to `RestClientsConfig`, while the nested config class would be `RestClientConfig` (for specific client configurations), similarly as GraphQL extension does (https://github.com/quarkusio/quarkus/pull/17973/files#diff-b292b26a4dd3fd549d35c404b15f61fc9e8923639e256774ab134a5f5797ba95).

radcortez commented on extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientConfigRootProvider.java:
```diff
@@ -0,0 +1,19 @@
+import javax.inject.Inject;
+
+@ApplicationScoped
+public class RestClientConfigRootProvider {
```
I wouldn't include this, because the config may be injected directly and is just adding an additional layer for little benefit.

radcortez commented on extensions/resteasy-classic/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientBase.java:
```diff
@@ -285,9 +374,29 @@ private void configureBaseUrl(RestClientBuilder builder) {
         return config.getOptionalValue(propertyName, type);
     }
 
+    private <PropertyType> Optional<PropertyType> getClientConfigPropertyWithFallback(
```
I think that any fallback / relocate of configuration properties should use the SR Config APIs for that. In that way, any lookup using the Config API will apply the relocate / fallback instead of just doing it here. You can check a few examples here:

https://smallrye.io/docs/smallrye-config/interceptors/interceptors.html#fallback-interceptor
https://github.com/quarkusio/quarkus/blob/abb7978101f98ebfc8eba201140395347f526137/extensions/smallrye-opentracing/runtime/src/main/java/io/quarkus/smallrye/opentracing/runtime/TracingConfigRelocateConfigSourceInterceptor.java

TomasHofman replied:
Done, see `RestClientFallbackConfigSourceInterceptor`.

TomasHofman commented:
@radcortez @FroMage thanks for reviews! I'm working on changes.

TomasHofman commented:
> Also, question: did you really implement that `@RegisterRestClient.configKey` defaults to the simple class name?

@FroMage good catch, I forgot about short names. They are implemented now, thanks!

TomasHofman commented:
Working on the test failures...

TomasHofman commented:
The CI issues are resolved now.

FroMage reviewed (changes requested):
LGTM, except the rest-client and rest-client-reactive documentation guides should be updated with the new recommended options, no?

TomasHofman commented:
I will look at the documentation shortly.

TomasHofman committed:
Quarkus config properties for MP REST Client - documentation

TomasHofman commented:
Rebased and added documentation changes to reference the new properties.

TomasHofman commented:
Investigating test failure.

TomasHofman commented:
The remaining test failure (LambdaDevServicesContinuousTestingTestCase) doesn't seem to be related to changes in this PR.

stuartwdouglas commented:
That is a known intermittent failure on windows, I will look into it today.

stuartwdouglas commented:
Fix for  the intermittent failure is here:  https://github.com/quarkusio/quarkus/pull/20012

FroMage reviewed (approved):
LGTM

radcortez commented on docs/src/main/asciidoc/rest-client.adoc:
```diff
@@ -191,8 +189,8 @@ The name of the property needs to follow a certain convention which is best disp
 # Your configuration properties
-org.acme.rest.client.CountriesService/mp-rest/url=https://restcountries.eu/rest # // <1>
-org.acme.rest.client.CountriesService/mp-rest/scope=javax.inject.Singleton # // <2>
+quarkus.rest-client."org.acme.rest.client.CountriesService".url=https://restcountries.eu/rest # // <1>
```
I'm ok if we recommend using the quarkus properties instead, but the old ones are still valid, so I think we should add a note to clarify which ones have priority if you have both set (shouldn't happen, but still....)

TomasHofman replied:
I added a short note stating that the "standard MicroProfile Rest Client notation" can be used too, while Quarkus notation takes a precedence. Thanks @radcortez!

radcortez reviewed (approved):
Just added a small comment in the documentation, but nothing major.

michalszynkiewicz commented on docs/src/main/asciidoc/config-yaml.adoc:
```diff
@@ -55,10 +55,10 @@ quarkus:
-      CountriesService/mp-rest/url: https://restcountries.eu/rest
+quarkus:
+  rest-client:
+    "org.acme.restclient.CountriesService":
```
do we need quotes here?

TomasHofman replied:
No it works without quotes too.

TomasHofman replied:
I removed the quotes from the example, I think it's better to keep it simpler...

TomasHofman committed:
Quarkus config properties for MP REST Client - documentation

* removing quotes around class name in a json config,
* adding note about standard MP notation being supported.

TomasHofman commented on docs/src/main/asciidoc/rest-client.adoc:
```diff
@@ -202,6 +200,20 @@ The default scope can also be defined on the interface.
 
 Note that `org.acme.rest.client.CountriesService` _must_ match the fully qualified name of the `CountriesService` interface we created in the previous section.
 
+[NOTE]
+====
+The standard MicroProfile Rest Client properties notation can also be used to configure the client:
+
+[source,properties]
+----
+org.acme.rest.client.CountriesService/mp-rest/url=https://restcountries.eu/rest
+org.acme.rest.client.CountriesService/mp-rest/scope=javax.inject.Singleton
+----
+
+If a property is specified via both the Quarkus notation and the MicroProfile notation, the Quarkus notation takes a precedence.
+====
+
+
```
Cc @radcortez

TomasHofman committed:
Use quarkus properties notation in exception message

TomasHofman committed:
Rest client configuration: fix tests

TomasHofman commented:
Added some test fixes after rebasing.

geoand commented:
Can you please rebase this to make sure it still passes CI?

Sorry it wasn't merged before...

geoand commented:
I actually rebased the PR myself.

TomasHofman commented:
Thanks! I will watch the results.
